### PR TITLE
Refactor Sessions Out

### PR DESCRIPTION
**Pull request by rudyjahchan**
This PR removes the concept of sessions being stored in sheets. Instead, when a collaborator connects to a board, their join message is relayed to the other connected collaborators. They reply with a "heartbeat" (which is also sent out every minute). It is then left up to the client side to maintain the lists of online/offline collaborators.

Notes:

* This does not complete the sheets work, so existing bugs and issues (such as the blur not working) still exist.
* No migration is included at this time, but at some point we should write a migration to pull out the sessions completely from the stored sheets.
* It does not save the heartbeat in the collaborator

### CODE REVIEW

File: assets/js/handler.coffee [hunk 1]
```diff
@@ -127,46 +132,46 @@ class boardroom.Handler
...
-    @board.userJoined message.userId
+    @board.userJoined message.user
+    @sendHeartbeat()
 
   onLeave: (message) =>
     @logger.debug 'onLeave'
-    @board.setOnlineUsers message.users
-    @board.userLeft message.userId
+    @board.userLeft message.user
+    @sendHeartbeat()
```
Comment (mwynholds) [inline assets/js/handler.coffee]: We discussed, as an optimization, only sending a heartbeat when we hear a "leave" message if the user who left is the same as the current user.

File: assets/js/models/board.coffee [hunk 2]
```diff
@@ -34,26 +39,35 @@ class boardroom.models.Board extends Backbone.Model
...
-          @onlineUsers().remove identity
-          @offlineUsers().push identity unless alreadyOffline(userId)
-
-  userJoined: (userId) =>
-
-  userLeft: (userId) =>
+  userJoined: (user) =>
+    {userId} = user
+    identity = new boardroom.models.UserIdentity(_.extend({}, user, {lastHeartbeatHeardAt: Date.now()}))
+    @collaborators().add(identity)
```
Comment (mwynholds) [inline assets/js/models/board.coffee]: A user who joins may already be a collaborator.  So I think we need to make sure to only add if the user doesn't already exist in the collection.

File: assets/js/models/board.coffee [hunk 3]
```diff
@@ -34,26 +39,35 @@ class boardroom.models.Board extends Backbone.Model
...
+  alreadyOffline: (userId) =>
+    @offlineUsers().find (identity) ->
+      identity.userId() == userId
+
+  offlineFlatlinedUserIdentities: =>
+    expiration = new Date(new Date().getTime() - FLATLINED_AGE)
+    flatLinedUsers = @onlineUsers().filter (user) -> user.lastHeartbeatHeardAt()? and user.lastHeartbeatHeardAt() < expiration
+    flatLinedUsers.forEach (user) => @userLeft(user.attributes)
+
+  onStatusChanged: =>
```
Comment (mwynholds) [inline assets/js/models/board.coffee]: I'm not sure what this is doing.  Can you explain?

File: lib/models/sheet.coffee [hunk 4]
```diff
@@ -34,46 +28,16 @@ SheetSchema.pre 'remove', (next) ->
...
-
-  offlinify: (socket) ->
-    session = _(@online).find (s) -> s.socket == socket
-    @online.remove session
-
-  heartbeat: (socket) ->
-    session = _(@online).find (s) -> s.socket == socket
-    session.heartbeat = Date.now()
-
   removeSessionsOlderThan: (millis) ->
```
Comment (mwynholds) [inline lib/models/sheet.coffee]: Can't we get rid of this as well?

File: lib/services/sockets/manager.coffee [hunk 5]
```diff
@@ -15,16 +15,14 @@ class Manager
...
-          collaborator = _.sample(sheet.get('online'))
-          if (collaborator)
-            room = board.id
-            redisClient.set("board_#{board.id}_collaborator_chosen", collaborator.user.id)
-            @io.to(room).emit 'screenshot', { userId: collaborator.user.id }
+      boardId = socket.rooms[1]
+      selectedScreenshooterSocket = _.sample(_.filter(io.sockets, (socket) -> socket.connected and !socket.disconnected and _.contains(socket.rooms, boardId)))
+      if selectedScreenshooterSocket?
+        redisClient.set("board_#{boardId}_collaborator_chosen", selectedScreenshooterSocket.id)
+        socket.emit 'screenshot', {}
```
Comment (mwynholds) [inline lib/services/sockets/manager.coffee]: Shouldn't this line be:

```
selectedScreenshooterSocket.emit 'screenshot', {}
```

File: lib/services/sockets/user.coffee [hunk 6]
```diff
@@ -1,56 +1,70 @@
...
+          return ( logger.error -> "Cannot find board: #{@boardId} (user: #{user.activeIdentity.profile()})" ) unless board?
+          @broadcastLeave(user, board, sheet)
+          name = user.activeIdentity.profile()
+          logger.info -> "#{name} has disconnected from board '#{board.name}'"
 
-  usersMessage: (user, board, sheet) ->
-    toIdentity = (s) -> s.user.activeIdentity
-    { userId: user.id, users: sheet.online.map toIdentity  }
+  registerHeartbeat: ->
+    @socket.on 'heartbeat', (message) =>
```
Comment (mwynholds) [inline lib/services/sockets/user.coffee]: I thought we were going to either:
a) just bail on all the logging so we could get rid of all the Mongo lookups here, or
b) at least pull the `@rebroadcastHeartbeat(user, board, sheet)` up to line 57 to it happens async with the Mongo lookups

File: lib/services/sockets/user.coffee [hunk 7]
```diff
@@ -60,7 +74,6 @@ class UserHandler
         next null, user, board
 
   leaveSheet: (session, board, sheet) ->
-    sheet.offlinify @socket.id
     sheet.save (err, _, num) ->
```
Comment (mwynholds) [inline lib/services/sockets/user.coffee]: Since we aren't `offlinify`-ing the socket, we don't need to save the sheet on line 77 at all.

File: package.json [hunk 8]
```diff
@@ -26,10 +26,8 @@
     "jade": "0",
     "jquery": "2",
     "less": "1",
-    "mongodb": "1.2",
```
Comment (mwynholds) [inline package.json]: I think we need the core mongo driver for the migrations.  I realize it comes with mongoose, but we don't use Mongoose for the migrations, and therefore I don't like to rely on it's dependencies to make them work.

File: lib/services/sockets/user.coffee [hunk 9]
```diff
@@ -1,56 +1,70 @@
...
+  registerHeartbeat: ->
+    @socket.on 'heartbeat', (message) =>
+      Sheet.findById @sheetId, (err, sheet) =>
+        @findUserAndBoard @userId, @boardId, (err, user, board) =>
+          return ( logger.error -> "Error finding user #{@userId} and board #{@boardId}: #{err}" ) if err?
+          return ( logger.error -> "Cannot find user: #{@userId}" ) unless user?
+          return ( logger.error -> "Cannot find board: #{@boardId} (user: #{user.activeIdentity.profile()})" ) unless board?
+          name = user.activeIdentity.profile()
+          logger.info -> "#{name} has sent heartbeat from board '#{board.name}'"
+          @rebroadcastHeartbeat(user, board, sheet)
```
Comment (mwynholds) [inline lib/services/sockets/user.coffee]: I thought we were going to either:
* just bail on all the logging so we could get rid of all the Mongo lookups here, or
* at least pull the `@rebroadcastHeartbeat(user, board, sheet)` up to line 57 to it happens async with the Mongo lookups

File: assets/js/models/board.coffee (hunk 2 above)
Comment (rudyjahchan) [inline assets/js/models/board.coffee]: @mwynholds That's why it uses an `add` instead of a `push` and other code declares `userId` as the `idAttribute` of `UserIdentity`. Backbone Collection's only `add` if there is no model with the same identifier already in the collection. This ensures only new collaborators are added.

File: assets/js/models/board.coffee (hunk 3 above)
Comment (rudyjahchan) [inline assets/js/models/board.coffee]: @mwynholds If the status changes to disconnect it moves all online users to offline.

File: lib/models/sheet.coffee (hunk 4 above)
Comment (rudyjahchan) [inline lib/models/sheet.coffee]: @mwynholds Good catch, will do!

File: lib/services/sockets/manager.coffee (hunk 5 above)
Comment (rudyjahchan) [inline lib/services/sockets/manager.coffee]: @mwynholds Good catch, will do!

File: lib/services/sockets/user.coffee (hunk 7 above)
Comment (rudyjahchan) [inline lib/services/sockets/user.coffee]: @mwynholds Good catch will do!

File: package.json (hunk 8 above)
Comment (rudyjahchan) [inline package.json]: @mwynholds Actually I am not sure why this change happened. Will remove.

File: lib/services/sockets/user.coffee (hunk 9 above)
Comment (rudyjahchan) [inline lib/services/sockets/user.coffee]: @mwynholds This is what I talked to you about EoD; that the message DOES need the full user and identity (as they get updated/inserted into online/offline), so it's not just there for logging purposes.

File: assets/js/handler.coffee (hunk 1 above)
Comment (rudyjahchan) [inline assets/js/handler.coffee]: @mwynholds Yeah, but I given all the changes, I wanted to make sure the general idea works before we start optimizing.

File: assets/js/models/board.coffee (hunk 2 above)
Comment (mwynholds) [inline assets/js/models/board.coffee]: Awesome.  I didn't know that about Backbone collections.  There are other places in the code, I think, that I am doing that check manually.  Good potential refactor.

File: lib/services/sockets/user.coffee (hunk 9 above)
Comment (mwynholds) [inline lib/services/sockets/user.coffee]: I wonder if we can just send all that info in the message, and then just pass it directly along.  Even when we wind up saving a timestamp in the db, we can rebroadcast the the heartbeat asynchronously, and we can (potentially) optimize the db access by throttling it.  (ie: maybe we only save each 10th heartbeat per user, or something similar)
